### ISSUE TITLE
Rework the way of changing current scene

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#921]
**Describe the project you are working on:**
Games made with Godot.

**Describe the problem or limitation you are having in your project:**
This is actually an issue I used to have, but now it impacts me less. Still, I've seen some users having this problem and it was generally agreed that the system should change.

Right now we have 2 methods to change scene: `change_scene` and `change_scene_to`. Their naming is already confusing for some people and there are a few problems with changing scenes that these methods don't solve.

First, sometimes you want to setup the new scene. Right now it's impossible, because you don't have access to the newly changed scene in any way (`current_scene` doesn't point to the new scene immediately). So if you want to pass some data from old scene to the new one, you need to use singleton.

Second, sometimes you want to keep the old scene, e.g. when you open in-game menu and you want to restore game state later. Currently `change_scene` will automatically free the old scene.

The current workaround for both is changing the scenes manually, i.e. accessing scene tree, replacing root and setting the `current_scene`. I noticed that I actually do this quite frequently in my projects, which makes regular `change_scene` quite pointless, because it's not flexible enough. It might be alright for some prototypes or learning projects, but there aren't many games for which the current workflow is enough. Also official docs propose that you have a "main node" and keep the game state as a child of your node. This makes SceneTree root and `change_scene` methods even more useless xd

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
The system that was proposed originally (I mean after my original issue and then discussion about the now-closed PR) is to have three methods:
```
change_scene_path(path, keep_old = false)
change_scene_packed(packed_scene, keep_old = false)
change_scene_node(node, keep_old = false)
```
The first one takes path to the scene resource, the second one takes a PackedScene and instances it and the third one uses an already existing node. Each of them has additional argument `keep_old`, which if true, makes the old scene not being freed automatically. That scene would be returned from the method. And if `keep_old` is false, the method would free the old scene and return null.

*Alternatively, the methods could return the new scene. It has the advantage that all the methods would give you its reference (for third method you have the reference anyways, because you provide it as argument), but the downside is that you'd need to keep the old scene yourself.*

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
What I said above, but in C++ bruh
There was an old PR that added something similar: https://github.com/godotengine/godot/pull/26833

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Well, it can be worked around quite easily:
```
var menu = preload("res://Menu.tscn").instance()
menu.game = self
get_node("/root").add_child(menu)
get_tree().current_scene = menu 
get_node("/root").remove_child(self)
```
But this code is not obvious to figure out for beginners in case they want to do either 1) or 2). Also this isn't really rarely used IMO.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because it would replace the current `change_scene/to` methods.

**Comment by alexzheng** [comment]
change_scene method return the new scene instance immediately would be enough.

**Comment by KoBeWi** [comment]
But that means you can't do scene initialization inside `_ready()`, because it's called before you get access to the scene.

**Comment by alexzheng** [comment]
yes, my_setup would be called after _ready.

To do setup before added to scene tree, one way is pass the node instance, another is pass parameters  to change_scene and do setup in the _init method of the scene.

**Comment by Calinou** [comment]
> Right now we have 2 methods to change scene: `change_scene` and` change_scene_to`. Their naming is already confusing for some people and there are a few problems with changing scenes that these methods don't solve.

This is taken care of by https://github.com/godotengine/godot/pull/38209.

**Comment by dalexeev** [comment]
> change_scene_node(node, keep_old = false)

Yes, this method would be useful. It would be possible (only in 4.0, since there is a cyclical dependency):

```
class_name Scene
extends Node

const SCENES_PATH = "res://scenes/%s.tscn"

var args := {}

func goto(path: String, args := {}) -> void:
    var next := load(SCENES_PATH % path).instance() as Scene
    next.args = args
    get_tree().change_scene_node(next)
```
